### Remove punctuation from string
 Made by Giang 

### Option 1 Regular expression

In [18]:
#Option 1 
import re
s = 'string with "punctuation" inside of it! Does this work? I hope so.'
fline = re.sub('['+string.punctuation+']', '', s)
fline

'string with punctuation inside of it Does this work I hope so'

###  Option 2 : string.maketrans() + translate()
- Python provides a constant called string.punctuation that provides a great list of punctuation characters. For example:
print(string.punctuation)
!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~
- Python offers a function called translate() that will map one set of characters to another.

- We can use the function maketrans() to create a mapping table. We can create an empty mapping table, but the third argument of this function allows us to list all of the characters to remove during the translation process. For example:

table = str.maketrans('', '', string.punctuation)

We can put all of this together, load the text file, split it into words by white space, then translate each word to remove the punctuation.

In [22]:
# Option 2.1
import string
def remove_punctuation(x):
    table = str.maketrans({key: None for key in string.punctuation})
    return x.translate(table)
s = 'string with "punctuation" inside of it! Does this work? I hope so.'
print(remove_punctuation(s))

string with punctuation inside of it Does this work I hope so


In [25]:
# or Option 2.2 
def remove_punctuation1(x):
    table = str.maketrans('','', string.punctuation)
    return x.translate(table)
print(remove_punctuation1(s))

string with punctuation inside of it Does this work I hope so


In [29]:
# Option 2.3
import string
table = str.maketrans('', '', string.punctuation)
stripped = [w.translate(table) for w in s]
print(s)

string with "punctuation" inside of it! Does this work? I hope so.


### The str.maketrans() 
This method returns a mapping table for translation usable for translate() method.The syntax of maketrans() method is: 
####  string.maketrans(x[, y[, z]])  

The maketrans() method takes 3 parameters:
- x - If only one argument is supplied, it must be a dictionary.
The dictionary should contain 1-to-1 mapping from a single character string to its translation OR a unicode number (97 for 'a') to its translation.

- y - If two arguments are passed, it must be two strings with equal length.
Each character in the first string is a replacement to its corresponding index in the second string.

- z - If three arguments are passed, each character in the third argument is mapped to None.

  Here, y and z are optional arguments.
    

In [21]:
# https://www.programiz.com/python-programming/methods/string/maketrans
#Example1
dict = {"a": "123", "b": "456", "c": "789"}
string = "abc"
print(string.maketrans(dict))
#Example 2
dict = {97: "123", 98: "456", 99: "789"}
string = "abc"
print(string.maketrans(dict))


{97: '123', 98: '456', 99: '789'}
{97: '123', 98: '456', 99: '789'}


Here, a dictionary dict is defined. It contains a mapping of characters a,b and c to 123, 456 and 789 respectively.

maketrans() creates a mapping of the character's Unicode ordinal to its corresponding translation.

So, 97 ('a') is mapped to '123', 98 'b' to 456 and 99 'c' to 789. This can be demonstrated from the output of both dictionaries.


In [20]:
# Example 2 : Translation table using two strings with maketrans()
# first string
firstString = "abc"
secondString = "def"
string = "abc"
print(string.maketrans(firstString, secondString))



{97: 100, 98: 101, 99: 102}


Here first, two strings of equal length "abc" and "def" are defined. And the corresponding translation is created.

Printing only the first translation gives you a 1-to-1 mapping to each character's Unicode ordinal in firstString to the same indexed character on the secondString.

In this case, 97 ('a') is mapped to 100 ('d'), 98 ('b') to 101 ('e') and 99 ('c') to 102 ('f').

In [13]:
# Example 2 Continue : example dictionary
firstString = "abc"
secondString = "defghi"
string = "abc"
print(string.maketrans(firstString, secondString))

ValueError: the first two maketrans arguments must have equal length

Trying to create a translation table for unequal length strings raises a ValueError exception indicating that the strings must have equal length.

In [10]:
# Example 3 : Translational table with removable string with maketrans()

firstString = "abc"
secondString = "def"
thirdString = "abd"
string = "abc"
print(string.maketrans(firstString, secondString, thirdString))

{97: None, 98: None, 99: 102, 100: None}


Here, first the mapping between the two strings firstString and secondString are created.

Then, the third argument thirdString resets the mapping of each character in it to None and also creates new mapping for non-existent characters.

In this case, thirdString resets the mapping of 97 ('a') and 98 ('b') to None, and also creates a new mapping for 100 ('d') mapped to None.

### The str.translate() method
returns a string where each character is mapped to its corresponding character as per the translation table.

In [1]:
# first string
firstString = "abc"
secondString = "ghi"
thirdString = "ab"

string = "abcdef"
print("Original string:", string)

translation = string.maketrans(firstString, secondString, thirdString)

# translate string
print("Translated string:", string.translate(translation))

Original string: abcdef
Translated string: idef


Here, the translation mapping translation contains the mapping from a, b and c to g, h and i respectively.

But, the removal string thirdString resets the mapping to a and b to None.

So, when the string is translated using translate(), a and b are removed, and c is replaced i outputting idef.

In [2]:
# translation table - a dictionary
translation = {97: None, 98: None, 99: 105}

string = "abcdef"
print("Original string:", string)

# translate string
print("Translated string:", string.translate(translation))

Original string: abcdef
Translated string: idef


Here, we don't create a translation table from maketrans() but, we manually create the mapping dictionary translation.

This translation is then used to translate the string string to get the same output as the previous example.

In [27]:
firstString = "agk"
secondString = "123"
x= str.maketrans(firstString, secondString)
x
s= 'The cows and the cat are kute'
print(s.translate(x))

The cows 1nd the c1t 1re 3ute


In [31]:
import string
s= 'The cows and the cat are kute&^^&$#@'
x=str.maketrans('abc','123',string.punctuation)
print(s.translate(x))

The 3ows 1nd the 31t 1re kute


In [ ]:
#Old version

####  An easier way to understand:
 the syntax for maketrans() method −

#### str.maketrans(intab, outtab)

Parameters
intab − This is the string having actual characters.

outtab − This is the string having corresponding mapping character.

Return Value
This method returns a translate table to be used translate() function.

#### Str.translate()
It returns a copy of the string in which all characters have been translated using table (constructed with the maketrans() function in the string module), optionally deleting all characters found in the string deletechars.

Following is the syntax for translate() method −

#### str.translate(table[, deletechars]);

Parameters

table − You can use the maketrans() helper function in the string module to create a translation table.

deletechars − The list of characters to be removed from the source string.

In [8]:
intab = "aeiou"
outtab = "12345"
trantab = str.maketrans(intab, outtab)

str = "this is string example....wow!!!"
print (str.translate(trantab))
#Under this, every vowel in a string is replaced by its vowel position −

th3s 3s str3ng 2x1mpl2....w4w!!!


In [18]:
 # Required to call maketrans function.

intab = "aeiou"
outtab = "12345"
trantab = str. maketrans(intab, outtab,'xm')

str = "this is string example....wow!!!";
print (str.translate(trantab))
#Following is the example to delete 'x' and 'm' characters from the string −

th3s 3s str3ng 21pl2....w4w!!!
